src/WUWidget.ts

```diff
@@ -0,0 +1,295 @@
+                            .widgetClassID(result[0].classid)
+                            .dataSource(result[0].datasource || this._metaResult.wuid())
+                            .resultName(result[0].resultname)
+                            .mappings(result[0].mappings)
+                            .filteredBy(result[0].filteredby)
+                            .properties(result[0].properties))
+                        .filteredBy(result[0].filteredby)
+                        .properties(result[0].properties)
+                        ;
+                }
+            }
+            return this.resolveWidget().then(() => this);
+        });
+    }
+
+    refreshData(selections = {}, maxRowCount: number = -1) {
+        const widget = this.widget();
+        if (widget instanceof Graph) {
+            return Promise.all([this.verticesMeta().refreshData(selections, maxRowCount), this.edgesMeta().refreshData({}, -1)]).then(results => {
+                return this.refreshGraph(widget, results[0].data, this.verticesMeta().properties(), results[1].data, this.edgesMeta().properties());
+            });
+        } else {
+            return this.widgetMeta().refreshData(selections, maxRowCount).then(results => {
+                return this.refreshWidget(widget, results);
+            });
+        }
+    }
+
+    refreshWidget(widget: Widget, result: { columns: any[]; data: any[]; }) {
+        widget
+            .columns(result.columns)
+            .data(result.data)
+            .lazyRender()
+            ;
+    }
+
+    refreshGraph(widget: Graph, vData: object[], vProps, eData: object[], eProps) {
+        const vertexMap: { [id: string]: Vertex } = {};
+        const vertices: Vertex[] = vData.map(row => {
+            const vertex = new Vertex()
+                .text(row[1])
+                ;
+            if (row[2]) {
+                vertex.faChar(row[2]);
+            } else {
+                vertex.icon_diameter(0);
+            }
+            vProps.forEach(row => {
+                if (typeof vertex[row.key] === "function") {
+                    vertex[row.key](row.value);
+                }
+            });
+            vertexMap[row[0]] = vertex;
+            return vertex;
+        });
+        const edges: Edge[] = eData.map(row => {
+            const source = vertexMap[row[0]];
+            const target = vertexMap[row[1]];
+            if (source && target) {
+                const edge = new Edge()
+                    .sourceVertex(source)
+                    .targetVertex(target)
+                    .text(row[2])
+                    ;
+                if (row[2]) {
+                    edge.text(row[2]);
+                }
+                eProps.forEach(row => {
+                    if (typeof edge[row.key] === "function") {
+                        edge[row.key](row.value);
+                    }
+                });
+                return edge;
+            }
```

Comment:
Logging a meaningful error in an else condition here could potentially prevent headaches.

Reply:
Change made as requested